### ISSUE TITLE
Exchange mailbox pool: Error with code 302

### ISSUE THREAD

**Issue by AlexTalker** [thunderbird/thunderbird-android#1772]
### Expected behavior
I wanted to use my work mailbox with k9-mail.
I guessed it will work following wiki page about this feature.

### Actual behavior
After successful authorization, inbox can't pool anything and fail with error 'IOException: Error with code 302 during request processing: HTTP/1.1 302 Redirect'.
I looked at the code by trace and seems like problem is here https://github.com/k9mail/k-9/blob/78d3b011e658fcddd3c23d284b32a16e49f467e3/k9mail-library/src/main/java/com/fsck/k9/mail/store/webdav/WebDavStore.java#L1026
I don't understand why status 300, 301 & 302 count as error since its just a forward, might be best just follow pages which server offers?

Trace:
```
K9-Mail version: 5.010
Device make: Micromax
Device model: Micromax Q415
Android version: 5.1.1

com.fsck.k9.mail.MessagingException: IOException
	at com.fsck.k9.mail.store.WebDavStore.sendRequest(WebDavStore.java:1125)
	at com.fsck.k9.mail.store.WebDavStore.processRequest(WebDavStore.java:1167)
	at com.fsck.k9.mail.store.WebDavStore.processRequest(WebDavStore.java:1142)
	at com.fsck.k9.mail.store.WebDavStore.access$500(WebDavStore.java:58)
	at com.fsck.k9.mail.store.WebDavStore$WebDavFolder.getMessageCount(WebDavStore.java:1361)
	at com.fsck.k9.mail.store.WebDavStore$WebDavFolder.getMessageCount(WebDavStore.java:1375)
	at com.fsck.k9.controller.MessagingController.synchronizeMailboxSynchronous(MessagingController.java:1053)
	at com.fsck.k9.controller.MessagingController.access$400(MessagingController.java:111)
	at
com.fsck.k9.controller.MessagingController$8.run(MessagingController.java:934)
	at com.fsck.k9.controller.MessagingController.run(MessagingController.java:435)
	at java.lang.Thread.run(Thread.java:818)
Caused by: java.io.IOException: Error with code 302 during request processing: HTTP/1.1 302 Redirect
	at com.fsck.k9.mail.store.WebDavStore.sendRequest(WebDavStore.java:1113)
	... 10 more
```

### Steps to reproduce
Create account WebDAV follow alternate way as decribes here https://github.com/k9mail/k-9/wiki/K9MailAndExchange
Try to pool inbox.

### Environment
K-9 Mail version: 5.010

Android version: 5.1.1

Account type (IMAP, POP3, WebDAV/Exchange): WebDAV/Exchange

**Comment by philipwhiuk** [comment]
So unlike IMAP it's difficult to test WebDAV instances. In general you're trying to work against the Microsoft implementation which isn't so much documented as reverse engineered. It's too expensive to purchase Microsoft Exchange server licenses purely for testing purposes. 

In general the code is written and tested by developers using their own work Exchange servers. I made a number of fixes when my work was using a version of Exchange that supported WebDAV. That's no longer possible for me because we upgraded to a newer version of Exchange.

This works in the majority of cases, but exercising all the code-paths is hard.

That said, we don't follow re-directs in general because the redirect URL we are sent is generally not what we want. So a re-direct unexpected is an error case.

On line 902 you can see comments to this effect.

I think at minimum we want to make some changes to that line to allow us to know what URL we are being redirected from and too. Then we can try and handle it at some point.

Often a redirect actually means 'not logged in' and you're being redirected to the login page. It may be really that the authentication is wrong. But without knowing why, we can't know whether this is the case.

**Comment by AlexTalker** [comment]
@philipwhiuk can I somehow help then?
I would like if I'll have some free time 'cause I work mailbox on exchange and I like to use K9 instead of other solutions for android as email client.
I also will appreciate if you point me to some docs, kinda architecture or something like if its exists. Your project is interesting but seems like it'll be hard to learn in short time.

**Comment by philipwhiuk** [comment]
So the PR I just created improves the logging so the error message you got would tell us where the redirect was trying to send us. To help us resolve the issue you could:

* Install AndroidStudio
* checkout the webDavLogging branch I just created
* Install the debug app on your device (it's a different package name so you can have both installed)
* Re-create your account in the debug app
* Look at the new error logs.

If you wanted you could then look at the WebDAV store code and work out what we should be doing.

As for documentation, right now it's a bit fragmented. The GitHub wiki has quite a few notes and information but it's not that well organised: https://github.com/k9mail/k-9/wiki For plug-ins and so on I'm currently building out the website covering what features we support. https://k9mail.github.io/documentation/development.html

I agree an architecture page would be useful - I'll try and workout where that's best to go - it'll probably be on the website.

**Comment by AlexTalker** [comment]
@philipwhiuk so I spent a few hours and:
* My OWA seems 2010 :disappointed: probably I should guess that by copyright
* Just as wiki says, API is changed, so ever with success login I can't get anything by '<owa-path>/<user>/INBOX' or at least my request is wrong.
* I wanna try debug it a little more, but seems like comparing by <owa-path>/<username> is wrong since on success 20120 wants redirect me to just <owa-path>
* Follow all 302 is bad idea, need stop as soon as cookie with us(which seems current code does), 'cause otherwise I got info that my browser is not support js and logon page again.
* I wanna play a little more and even implement basic support of 2010.

Thank you for useful links.
By the way, bug seems my problem since I specified wrong owa path(/Exchange tried redirect to /OWA).
